src/source/Ice/NatBehaviorDiscovery.c

```diff
@@ -0,0 +1,404 @@
+    }
+    MUTEX_UNLOCK(pNatTestData->lock);
+    CVAR_SIGNAL(pNatTestData->cvar);
+
+CleanUp:
+
+    return STATUS_SUCCESS;
+}
+
+STATUS executeNatTest(PStunPacket bindingRequest, PKvsIpAddress pDestAddr, PSocketConnection pSocketConnection,
+                      UINT32 testIndex, PNatTestData pData, PStunPacket *ppTestReponse)
+{
+    PStunPacket testResponse = NULL;
+    UINT32 i, j;
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(bindingRequest != NULL && pDestAddr != NULL && pSocketConnection != NULL && pData != NULL && ppTestReponse != NULL, STATUS_NULL_ARG);
+
+    MEMSET(bindingRequest->header.transactionId, 0x00, STUN_TRANSACTION_ID_LEN);
+    putInt32((PINT32) bindingRequest->header.transactionId, testIndex);
+    for (i = 0; i < pData->bindingResponseCount; ++i) {
+        freeStunPacket(&pData->bindingResponseList[i]);
+    }
+    pData->bindingResponseCount = 0;
+
+    for (i = 0; testResponse == NULL && i < DEFAULT_NAT_TEST_MAX_BINDING_REQUEST_COUNT; ++i) {
+        iceUtilsSendStunPacket(bindingRequest, NULL, 0, pDestAddr, pSocketConnection, NULL, FALSE);
+        CVAR_WAIT(pData->cvar, pData->lock, 500 * HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+        if (pData->bindingResponseCount > 0) {
+            for (j = 0; j < pData->bindingResponseCount; ++j) {
+                if (getInt32(*(PINT32) pData->bindingResponseList[j]->header.transactionId) == testIndex) {
+                    testResponse = pData->bindingResponseList[j];
+                    break;
+                }
+            }
+        }
+    }
+
+CleanUp:
+
+    if (ppTestReponse != NULL) {
+        *ppTestReponse = testResponse;
+    }
+
+    CHK_LOG_ERR(retStatus);
+    return retStatus;
+}
+
+STATUS getMappAddressAttribute(PStunPacket pBindingResponse, PStunAttributeAddress *ppStunAttributeAddress)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    PStunAttributeAddress pStunAttributeAddress = NULL;
+
+    CHK(pBindingResponse != NULL && ppStunAttributeAddress != NULL, STATUS_NULL_ARG);
+
+    CHK_STATUS(getStunAttribute(pBindingResponse,
+                                STUN_ATTRIBUTE_TYPE_XOR_MAPPED_ADDRESS,
+                                (PStunAttributeHeader *) &pStunAttributeAddress));
+    if (pStunAttributeAddress == NULL) {
+        CHK_STATUS(getStunAttribute(pBindingResponse,
+                                    STUN_ATTRIBUTE_TYPE_MAPPED_ADDRESS,
+                                    (PStunAttributeHeader *) &pStunAttributeAddress));
+    }
+
+    CHK_ERR(pStunAttributeAddress != NULL, STATUS_INVALID_OPERATION,
+            "Expect binding response to have mapped address or xor mapped address attribute");
+
+CleanUp:
+
+    if (ppStunAttributeAddress != NULL) {
+        *ppStunAttributeAddress = pStunAttributeAddress;
+    }
+
+    CHK_LOG_ERR(retStatus);
+    return retStatus;
+}
+
+STATUS natMappingBehaviorTest(PIceServer pStunServer, PNatTestData data, PSocketConnection pSocketConnection,
```

Comment:
MINOR: like the other naming - natMappingBehaviorCheck?

Reply:
this mirrors the "test" described in the rfc: https://tools.ietf.org/html/rfc5780#section-4.3